comp/networkpath/npscheduler/npschedulerimpl/npscheduler.go

```diff
@@ -7,31 +7,324 @@
+
+	logger.Infof("New NpScheduler (workers=%d input_chan_size=%d pathtest_ttl=%s pathtest_interval=%s)",
+		workers,
+		pathtestInputChanSize,
+		pathtestTTL.String(),
+		pathtestInterval.String())
+
+	return &npSchedulerImpl{
+		enabled:     true,
+		epForwarder: epForwarder,
+		logger:      logger,
+
+		pathtestStore:       pathteststore.NewPathtestStore(pathtestTTL, pathtestInterval, logger),
+		pathtestInputChan:   make(chan *common.Pathtest, pathtestInputChanSize),
+		pathtestProcessChan: make(chan *pathteststore.PathtestContext, pathtestProcessChanSize),
+		flushInterval:       flushInterval,
+		workers:             workers,
+
+		metricSender: metricsender.NewMetricSenderStatsd(statsd.Client),
+
+		receivedPathtestCount:    atomic.NewUint64(0),
+		processedTracerouteCount: atomic.NewUint64(0),
+		TimeNowFn:                time.Now,
+
+		stopChan:      make(chan struct{}),
+		runDone:       make(chan struct{}),
+		flushLoopDone: make(chan struct{}),
+
+		statsdClient: statsd.Client,
+
+		runTraceroute: runTraceroute,
+	}
+}
+
+func (s *npSchedulerImpl) ScheduleConns(conns []*model.Connection) {
 	if !s.enabled {
 		return
 	}
-	// TODO: IMPLEMENTATION IN SEPARATE PR (to make PRs easier to review)
+	startTime := time.Now()
+	for _, conn := range conns {
+		// Only process outgoing traffic
+		if !shouldScheduleNetworkPathForConn(conn) {
+			continue
+		}
+		remoteAddr := conn.Raddr
+		remotePort := uint16(conn.Raddr.Port)
+		err := s.scheduleOne(remoteAddr.Ip, remotePort)
+		if err != nil {
+			s.logger.Errorf("Error scheduling pathtests: %s", err)
+		}
+	}
+
+	scheduleDuration := time.Since(startTime)
+	s.statsdClient.Gauge("datadog.network_path.scheduler.schedule_duration", scheduleDuration.Seconds(), nil, 1) //nolint:errcheck
 }
 
-func newNoopNpSchedulerImpl() *npSchedulerImpl {
-	return &npSchedulerImpl{
-		enabled: false,
+func shouldScheduleNetworkPathForConn(conn *model.Connection) bool {
+	return conn.Direction == model.ConnectionDirection_outgoing
+}
+
+// scheduleOne schedules pathtests.
+// It shouldn't block, if the input channel is full, an error is returned.
+func (s *npSchedulerImpl) scheduleOne(hostname string, port uint16) error {
+	if s.pathtestInputChan == nil {
+		return errors.New("no input channel, please check that network path is enabled")
+	}
+	s.logger.Debugf("Schedule traceroute for: hostname=%s port=%d", hostname, port)
+
+	if net.ParseIP(hostname).To4() == nil {
+		// TODO: IPv6 not supported yet
+		s.logger.Debugf("Only IPv4 is currently supported. Address not supported: %s", hostname)
+		return nil
 	}
+
+	ptest := &common.Pathtest{
+		Hostname: hostname,
+		Port:     port,
+	}
+	select {
+	case s.pathtestInputChan <- ptest:
+		return nil
+	default:
+		return fmt.Errorf("scheduler input channel is full (channel capacity is %d)", cap(s.pathtestInputChan))
+	}
+}
+func (s *npSchedulerImpl) start() error {
+	if s.running {
+		return errors.New("server already started")
```

Comment:
Usually this is a no-op (without an error) in this kind of situation.

Reply:
In the current implement, it's not expected for `start()` to be called multiple time, if it does there is probably an issue and I think it's good to fail and inspect why it's happening.

That said, I think in other situation, it might be preferred to use a no-op approach.